### PR TITLE
Initial commit of gcs modular file system plugin (vnvo2409)

### PR DESCRIPTION
**Pull request by yongtang**
This is the initial commit of gcs modular file system plugin authored
by @vnvo2409. This is the first step with code moved here.

Several things that needs to be done in follow up PRs:
1. The thread-related functions was using TensorFlow APIs. However, they are not exposed in TF 2.4 so we may need to convert them here first.
2. The testing has not been touched yet. It will be covered in follow up PRs.

This PR is a first step, as the content is pretty big so split into steps.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

File: tensorflow_io/core/plugins/gs/cleanup.h [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+//     }
+//   }
+//
+// You can use Cleanup<F> directly, instead of using MakeCleanup and auto,
+// but there's rarely a reason to do that.
+//
+// You can call 'release()' on a Cleanup object to cancel the cleanup.
+
+#ifndef TENSORFLOW_C_EXPERIMENTAL_FILESYSTEM_PLUGINS_GCS_CLEANUP_H_
+#define TENSORFLOW_C_EXPERIMENTAL_FILESYSTEM_PLUGINS_GCS_CLEANUP_H_
```
Comment (kvignesh1420) [inline tensorflow_io/core/plugins/gs/cleanup.h:43]: @yongtang the guards can be renamed.

**Comment by vnvo2409** [comment]
LGTM. Thank you !

File: tensorflow_io/core/plugins/gs/cleanup.h (hunk 1 above)
Comment (yongtang) [reply tensorflow_io/core/plugins/gs/cleanup.h]: @kvignesh1420 Thanks. The PR has been updated with guard renamed.

**Review by kvignesh1420 — APPROVED** [review]

**Comment by kvignesh1420** [comment]
@yongtang thank you for migrating the plugins. Also, credit to @vnvo2409 for writing the plugins.

**Pull request merged**